src/DFAMin.hs

```diff
@@ -1,218 +1,258 @@
+    IntMap.restrictKeys m s
 #else
-        preimage invMap = IntSet.unions . mapMaybe (`IntMap.lookup` invMap) . IntSet.toList
+    IntMap.filterWithKey (\k _ -> k `IntSet.member` s) m
 #endif
 
-        xs :: [EquivalenceClass]
-        xs =
-          [ x
-          | invMap <- IntMap.elems bigmap
-          , let x = preimage invMap a
-          , not (IntSet.null x)
-          ]
-
-        refineWith
-          :: EquivalenceClass -- preimage set that bisects the input equivalence class
-          -> EquivalenceClass -- input equivalence class
-          -> Maybe (EquivalenceClass, EquivalenceClass) -- refined equivalence class
-        refineWith x y =
-          if IntSet.null y1 || IntSet.null y2
-            then Nothing
-            else Just (y1, y2)
-          where
-            y1 = IntSet.intersection y x
-            y2 = IntSet.difference   y x
-
-        go0 (r,q) x = go1 r [] []
-          where
-            -- iterates over R
-            go1 []    r' q' = (r', go2 q q')
-            go1 (y:r) r' q' = case refineWith x y of
-              Nothing                       -> go1 r (y:r') q'
-              Just (y1, y2)
-                | IntSet.size y1 <= IntSet.size y2 -> go1 r (y2:r') (y1:q')
-                | otherwise                        -> go1 r (y1:r') (y2:q')
-
-            -- iterates over Q
-            go2 []    q' = q'
-            go2 (y:q) q' = case refineWith x y of
-              Nothing       -> go2 q (y:q')
-              Just (y1, y2) -> go2 q (y1:y2:q')
-
--- To pacify GHC 9.8's warning about 'head'
-headWithDefault :: a -> [a] -> a
-headWithDefault a []    = a
-headWithDefault _ (a:_) = a
+
+-- | Given two sets X and Y, compute their intersection and difference.
+-- Only returns both if both are non-empty, otherwise return neither.
+refine
+  :: EquivalenceClass
+  -> EquivalenceClass
+  -> Maybe (EquivalenceClass, EquivalenceClass)
+refine x y =
+  if IntSet.null intersection || IntSet.null difference
+    then Nothing
+    else Just (intersection, difference)
+  where
+    intersection = IntSet.intersection y x
+    difference   = IntSet.difference   y x
+
+
+-- | Given a DFA, compute all sets of equivalent states.
+--
+-- See Note [Hopcroft's Algorithm] for details.
+groupEquivalentStates :: forall a. Ord a => DFA Int a -> [EquivalenceClass]
+groupEquivalentStates dfa = outerLoop ([], initialSubsets dfa)
+  where
+    reverseTransitionCache :: [IntMap EquivalenceClass]
+    reverseTransitionCache = generateReverseTransitionCache dfa
+
+    -- While W isn't empty, pick an A from W, add it to R
+    -- and iterate on X for each c in ∑.
+    outerLoop :: ([EquivalenceClass], [EquivalenceClass]) -> [EquivalenceClass]
+    outerLoop (r,  []) = r
+    outerLoop (r, a:w) = outerLoop $ List.foldl' refineWithX (a:r,w) $ do
+      allPreviousStates <- reverseTransitionCache
+      let x = IntSet.unions $ IntMap.elems $ restrictKeys allPreviousStates a
```

Comment:
Could it be just `fold`?
```suggestion
      let x = fold $ restrictKeys allPreviousStates a
```

Reply:
Let me try, I am just unsure about older versions of GHC.

Reply:
Done. I was concerned due to `IntSet.unions` being explicitly defined on list in older versions of `containers` (which is why i had an explicit call to `IntMap.elems`), but it builds. :+1: